frontend/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -43,6 +43,69 @@ export interface Props {
   allowHTML: boolean
 }
 
+function createAnchorFromText(text: string | null): string {
+  const newAnchor = text
+    ?.toLowerCase()
+    .split(/[^A-Za-z0-9]/)
+    .filter(Boolean)
+    .join("-")
+  return newAnchor || ""
+}
+
+function HeadingWithAnchor({
+  tag,
+  anchor: propsAnchor,
+  children,
+}: any): ReactElement {
+  const [anchor, setAnchor] = React.useState<string | null>(null)
+  const ref = React.useRef<HTMLElement>(null)
+
+  React.useEffect(() => {
+    if (anchor === null && ref.current !== null) {
+      const newAnchor =
+        propsAnchor || createAnchorFromText(ref.current.textContent)
+      if (newAnchor) {
+        setAnchor(newAnchor)
+        if (window.location.hash.slice(1) === newAnchor) {
+          ref.current.scrollIntoView(true)
+        }
+      }
+    }
+  }, [ref.current])
+
+  return React.createElement(
+    tag,
+    { ref },
+    <>
+      {anchor && <a data-anchor={anchor} />}
+      {children}
+    </>
```

Comment:
Could be id or name on the tag instead? See below for more

Reply:
See https://github.com/streamlit/streamlit/pull/2513#discussion_r548162142